File: download.py
import requests
import os
import sys
import time
from contextlib import closing
from fake_useragent import UserAgent

API_URL = "https://web.archive.org/cdx/search/cdx?url=*."
BASE_URL = "http://web.archive.org/web/"
useragent = UserAgent()

def getUrls(data, domain, timeframe):
    """
    Return a set of wayback URLs
    """
    wayback_urls = set()
    for record in data:
        items = record.split(' ')
        try:
            savedpage = items[2]
        except Exception:
            continue
        url = savedpage
        timestamp = items[1]
        if str(timestamp.strip()).startswith(timeframe):
            wayback_url = BASE_URL + timestamp + "/" + url
            wayback_urls.add(wayback_url)
    return wayback_urls

def download(session, savePath, url):
    """
    Download the webpage using requests with a single session and save it in the specified path.
    """
    noSlash = url.rstrip('/').replace('/', '£').replace(":", "!!!").replace("?", "§§")
    filename = noSlash + ".txt" if not noSlash.endswith('.txt') else noSlash
    output = os.path.join(savePath, filename)
     
    if len(filename) <= 255 and not os.path.exists(os.path.join(savePath, filename)):
        with closing(session.get(url, stream=True)) as response:
            if response.status_code == 200:
                with open(output, 'wb') as f:
                    for chunk in response.iter_content(chunk_size=1024):
                        f.write(chunk)
                    print("Downloaded:",url)
            else:
                print(f"Failed to download URL: {url} with status code: {response.status_code}")
    else:
        if len(filename) > 255:
            print("Skipped - filename too long", url)
        if os.path.exists(os.path.join(savePath, filename)):
            print("Skipped (file already existing):",url)

def main():
    domain = input("Type the target domain: ")
    timeStamp = str(input("Specify a timestamp, ex: yyyymmdd, but also: yyyymm. Type '2' or '1' > to download everything for the years past 20** or 19**): "))

    localDir = path = os.path.dirname(os.path.abspath(__file__))
    savePath = os.path.join(localDir, domain)

    try:
        os.mkdir(savePath)
    except FileExistsError:
        print("Resuming download")
        pass

    history = requests.get(API_URL + domain).text.splitlines()
    waybackurls = getUrls(history, domain, timeStamp)

    print("Preparing to download {} pages".format(str(len(waybackurls))))
    time.sleep(2) #try to reduce (or eliminate) this sleep time, but at your own risk of being blocked.

    with requests.Session() as session:
        session.headers = {'user-agent':useragent.chrome}
        for url in waybackurls:
            try:
                download(session, savePath, url)
            except Exception as e:
                print("Error downloading:",url)
                continue

if __name__ == "__main__":
    main()

File: download.go
package main

import (
	"fmt"
	"io"
	"log"
	"math/rand"
	"net/http"
	"os"
	"runtime"
	"strings"
	"sync"
	"time"
)

var path string
var targetDomain string
var timeStamp string
var chunksize int

func subslice(slice []string, chunkSize int) [][]string {
	var chunks [][]string
	for {
		if len(slice) == 0 {
			break
		}
		if len(slice) < chunkSize {
			chunkSize = len(slice)
		}
		chunks = append(chunks, slice[0:chunkSize])
		slice = slice[chunkSize:]
	}
	return chunks
}

func getPage(url string) string {
	var response string
	for n := 0; n <= 2; n++ {
		client := http.Client{Timeout: time.Duration(10) * time.Second}
		resp, err := client.Get(url)
		if err != nil {
			fmt.Println("Get Error:", url)
		} else {
			if resp.StatusCode == 200 {
				response_, _ := io.ReadAll(resp.Body)
				response = string(response_)
				resp.Body.Close()
				break
			} else {
				fmt.Println("Page not available:", url)
				response = "page not available"
				resp.Body.Close()
				break
			}
		}
	}
	return response
}

func pathExists(Path string) bool {
	_, error := os.Stat(Path)
	if os.IsNotExist(error) {
		return false
	} else {
		return true
	}
}

func setDomain() {
	path, _ = os.Getwd()
	fmt.Print("Specify the target domain (only lowercase): ")
	fmt.Scanln(&targetDomain)
	fmt.Print(
		"Specify timestamp in the format:'yyyymmdd' (also: 'yyyy' > download only a specific year; 'yyyymm' > year and month; '2' > everything): ",
	)
	fmt.Scanln(&timeStamp)
}

func createDir() {
	fmt.Println("Saving data in:", path+"/"+targetDomain)
	err_ := os.Mkdir(path+"/"+targetDomain, 0777)
	if err_ != nil {
		log.Fatal(err_)
	}
}

func saveFiles(array_ []string, waitgroup *sync.WaitGroup) {
	defer waitgroup.Done()
	for _, url := range array_ {
		urlstring_ := strings.Replace(url, "/", "£", -1)
		urlstring__ := strings.Replace(urlstring_, ":", "!!!", -1)
		urlstring := strings.Replace(urlstring__, "?", "§§", -1)
		file_name_check := urlstring + ".txt"
		pathToFile := path + "/" + targetDomain + "/" + file_name_check
		if pathExists(pathToFile) == false {
			if len(url) < 255 {
				content := getPage(url)
				if content != "page not available" {
					file, err := os.Create(pathToFile)
					if err != nil {
						fmt.Println(err)
					}
					file.WriteString(content)
					file.Close()
					fmt.Println("Done:", url)
				}
				frame := time.Duration(rand.Intn(100))
				time.Sleep(time.Millisecond * frame)
			}
		} else {
			fmt.Println("Skipping:", url)
		}
	}
}

func getHistory() []string {
	API_URL := "https://web.archive.org/cdx/search/cdx?url=*."
	query_url := API_URL + targetDomain
	client := http.Client{Timeout: time.Duration(50) * time.Second}
	response, err := client.Get(query_url)
	if err != nil {
		panic("can't download history for the specified domain ")
	}
	body, _ := io.ReadAll(response.Body)
	content := string(body)
	lines := strings.Split(string(content), "\n")
	BASE_URL := "http://web.archive.org/web/"
	var waybackurls []string
	for _, line := range lines {
		if len(line) > 0 {
			data := strings.Split(line, " ")
			savedpage := strings.Split(data[0], ")/")[1]
			url := targetDomain + "/"+ savedpage
			timestamp := string(data[1])
			if strings.HasPrefix(timestamp, timeStamp) {
				wayback_url := BASE_URL + timestamp + "/" + url
				waybackurls = append(waybackurls, wayback_url)
			}
		}
	}
	return waybackurls
}

func start(function func()) {
	var start_notice string
	if function != nil {
		function()
		start_notice = "Starting new download"
	} else {
		start_notice = "Resuming download"
	}
	history := getHistory()
	fmt.Println("Number of pages saved by Archive: ")
	fmt.Println(len(history))
	if len(history) >= 10 {
		chunksize = len(history) / 10
	} else {
		chunksize = len(history)
	}
	arrayOfArrays := subslice(history, chunksize)
	runtime.GOMAXPROCS(runtime.NumCPU())
	var waitgroup sync.WaitGroup
	waitgroup.Add(len(arrayOfArrays))
	fmt.Println(start_notice)
	for _, array := range arrayOfArrays {
		time.Sleep(time.Millisecond * 50)
		go saveFiles(array, &waitgroup)
	}
	waitgroup.Wait()
	fmt.Println("Download completed")
}

func main() {
	setDomain()
	if pathExists(path+"/"+targetDomain) == false {
		start(createDir)
	} else {
		start(nil)
	}
}
